gensim/models/keyedvectors.py

```diff
@@ -154,6 +154,84 @@ def get_vector(self, entity):
         else:
             raise KeyError("'%s' not in vocabulary" % entity)
 
+    def add_entity(self, entity, weights, replace=False):
+        """Add entity vector in a manual way.
+        If `entity` is already in the vocabulary, old vector is keeped unless `replace` flag is True.
+
+        Parameters
+        ----------
+        entity : str
+            Entity specified by string tag.
+        weights : np.array
+            1D numpy array with shape (`vector_size`,)
+        replace: bool, optional
+            Boolean flag indicating whether to replace old vector if entity is already in the vocabulary.
+            Default, False, means that old vector is keeped.
+        """
+        self.add_entities([entity], weights.reshape(1, -1), replace=replace)
+
+    def add_entities(self, entities, weights, replace=False):
+        """Add entities and theirs vectors in a manual way.
+        If some entity is already in the vocabulary, old vector is keeped unless `replace` flag is True.
+
+        Parameters
+        ----------
+        entities : list of str
+            Entities specified by string tags.
+        weights: {list of numpy.ndarray, numpy.ndarray}
+            List of 1D np.array vectors or 2D np.array of vectors.
+        replace: bool, optional
+            Flag indicating whether to replace vectors for entities which are already in the vocabulary,
+            if True - replace vectors, otherwise - keep old vectors.
+        """
+        if isinstance(weights, list):
+            weights = np.array(weights)
+
+        in_vocab_mask = np.zeros(len(entities), dtype=np.bool)
+        in_vocab_idxs = []
+        out_vocab_entities = []
+
+        for idx, entity in zip(range(len(entities)), entities):
+            if entity in self.vocab:
+                in_vocab_mask[idx] = True
+                in_vocab_idxs.append(self.vocab[entity].index)
+            else:
+                out_vocab_entities.append(entity)
+
+        # add new entities to the vocab
+        for entity in out_vocab_entities:
+            entity_id = len(self.vocab)
+            self.vocab[entity] = Vocab(index=entity_id, count=1)
+            self.index2entity.append(entity)
+
+        # add vectors for new entities
+        if len(self.vectors) == 0:
+            self.vectors = weights[~in_vocab_mask]
+        else:
+            self.vectors = vstack((self.vectors, weights[~in_vocab_mask]))
```

Comment:
Might this line work even in the case where `len(self.vectors)==0`, making the check/branch unnecessary?

Reply:
I think it's not obvious how to do that, because when empty `KeyedVectors` object is created, `self.vectors = []` is true. In that case, we can't use `vstack(([], weights[~in_vocab_mask]))` and `ValueError: all the input array dimensions except for the concatenation axis must match exactly` is raised.

Reply:
Is it possible for an empty `KeyedVectors` to have a `self.vectors` that is already a proper-dimensioned `(0, vector_size)` empty ndarray? (Not sure myself, but would simplify things in later places like this.)